### PR TITLE
Feature: Implement Graceful shutdown 

### PR DESCRIPTION
**Pull request by werniq**
This PR introduces graceful shutdown functionality to the server. It ensures that the server properly handles shutdown signals (`SIGINT`, `SIGTERM`), stops accepting new requests, and allows ongoing requests to complete before shutting down. The update includes:
- A new `Shutdown` method to gracefully stop the server with a timeout.
- Signal handling in the `Start` and `Run` methods to listen for shutdown signals and trigger the graceful shutdown process.

This change enhances the server's stability by ensuring a smooth and controlled shutdown process.

### CODE REVIEW

File: cmd/proxy/server/server.go [hunk 1]
```diff
@@ -95,18 +94,47 @@ func Run(ctx context.Context, proxyServerPort, visualizeServerPort string, yamlC
...
+	go func() {
+		if err := proxyServer.Start(); err != nil && !errors.Is(err, http.ErrServerClosed) {
+			log.Fatalf("Proxy server encountered an error: %v\n", err)
+		}
+	}()
+
+	select {
+	case sig := <-signalChan:
+		log.Printf("Received shutdown signal: %s, shutting down gracefully...", sig)
+		cancel()
```
Comment (werniq) [inline cmd/proxy/server/server.go:116]: As you can see here, when we receive signal into signalChan, we call cancel func, which closes our context, and 'graceful-shutdowning' process begins.

File: go.mod [hunk 2]
```diff
@@ -4,29 +4,31 @@ go 1.22.0
 
 require (
 	github.com/Netflix/go-env v0.0.0-20220526054621-78278af1949d
+	github.com/cdipaolo/goml v0.0.0-20220715001353-00e0c845ae1c
 	github.com/corazawaf/libinjection-go v0.1.3
 	github.com/emirpasic/gods v1.18.1
 	github.com/goccy/go-yaml v1.11.3
```
Comment (werniq) [inline go.mod:10]: Had to run go mod tidy here, seems we have some unused imports here

**Comment by werniq** [comment]
@cebilon123 while repairing CI/CD I've noticed that some tests took really long. After some investigation, I've reached to `rule` package. I'm not sure, but I can't see any usage of it in the Waffle. 

Is it needed?

File: Taskfile.yaml [hunk 3]
```diff
@@ -1,5 +1,8 @@
 version: '3'
 
+env:
+  IMAGE_NAME: '{{ .IMAGE_NAME | default "qniw984/waffle:1.0.0" }}'
```
Comment (thegodenage) [inline Taskfile.yaml:4]: Why qniw984 tho? 😄

File: cmd/proxy/main.go [hunk 4]
```diff
@@ -23,8 +27,25 @@ func main() {
...
+
+	signal.Notify(quit, syscall.SIGINT, syscall.SIGTERM)
+
+	go func() {
+		if err := server.Run(proxyServerPort, visualizeServerPort, yamlConfigBytes, certificates); err != nil {
+			log.Fatalln(err)
+		}
+	}()
+
+	log.Println("Server started on :8080")
```
Comment (thegodenage) [inline cmd/proxy/main.go:40]: I think it must be modified to the correct ports

File: pkg/permission/windows.go [hunk 5]
```diff
@@ -1,30 +1,30 @@
...
+//
+//	var showCmd int32 = 1 //SW_NORMAL
+//
+//	err := windows.ShellExecute(0, verbPtr, exePtr, argPtr, cwdPtr, showCmd)
+//	if err != nil {
+//		return fmt.Errorf("windows shell execute: %w", err)
+//	}
+//
+//	return nil
+//}
```
Comment (thegodenage) [inline pkg/permission/windows.go:30]: You can remove the whole file.

**Review by thegodenage — CHANGES_REQUESTED** [review]

**Comment by werniq** [comment]
@thegodenage what about `rule` package? It is not used in project, but tests from this package are failing. Can I remove it too?

File: Taskfile.yaml (hunk 3 above)
Comment (werniq) [reply Taskfile.yaml]: Changed to `<YOUR_USERNAME>`, so user can specify his/her username

File: cmd/proxy/main.go (hunk 4 above)
Comment (werniq) [reply cmd/proxy/main.go:40]: You mean like this:

image

File: pkg/permission/windows.go (hunk 5 above)
Comment (werniq) [reply pkg/permission/windows.go]: Removed

**Comment by werniq** [comment]
@thegodenage Finally, CI is blue 🙂

**Comment by thegodenage** [comment]
@werniq Hey sure, sorry for late response, have a lot of different stuff in the meantime regarding daytime job and apartment.

**Pull request merged**